base/src/main/java/proguard/evaluation/Variables.java

```diff
@@ -207,9 +213,14 @@ public Value load(int index)
         if (index < 0 ||
             index >= size)
         {
-            throw new IndexOutOfBoundsException("Variable index ["+index+"] out of bounds ["+size+"]");
+            throw new VariableInstructionEvaluationException("Variable index ["+index+"] out of bounds. There are "+size+" variables in this code attribute.",
+                    "You may be able to change the amount of variables you have through a field in codeAttribute, or use a smaller variable index");
         }
 
+        if (values[index] == null) {
+            throw new VariableInstructionEvaluationException("No value in variable slot "+index+".",
+                    "You might have forgotten to load the value first, or maybe you loaded it to another variable.");
+        }
```

Comment:
Maybe this can be extracted into a private method `checkBounds` or similar. Then it can be called from each of the `*load` methods like

```
    public IntegerValue iload(int index)
    {
         checkBounds(); // This will throw an exception if out of bounds.

         // At these point we know for sure that there is a value.

         Value value = values[index];

        // But is it of the correct type?
         
         if (value.computationalType() != IntegerValue.TYPE_INTEGER)
             throw new IncorrectVariableTypeException("....");
             
        return value.integerValue();
    }

```